File: backend/app/schemas/diagnosis.py
from datetime import datetime
from pydantic import BaseModel, Field


class PredictionItem(BaseModel):
    name: str
    confidence: float


class RecaptureCandidate(BaseModel):
    """혼동쌍 각 후보와, 판별부위에서 그 후보를 가리키는 시각 단서."""

    name: str
    cue: str


class RecaptureGuidance(BaseModel):
    """1차 진단이 혼동쌍에 낮은 마진으로 떨어졌을 때, 어느 부위를 다시 찍을지 안내."""

    region: str
    instruction: str
    margin: float
    candidates: list[RecaptureCandidate]


class PesticideSafety(BaseModel):
    phi_days: int | None = None
    reentry_hours: int | None = None


class PesticideInfo(BaseModel):
    name: str
    dilution: str | None = None
    usage: str | None = None
    safety: PesticideSafety = Field(default_factory=PesticideSafety)
    active_ingredient: str | None = None
    safety_notes: str | None = None


class DiagnoseResponse(BaseModel):
    id: str
    disease_name: str
    crop_type: str | None = None
    category: str = "disease"
    severity: str = "medium"
    confidence: float
    top_predictions: list[PredictionItem]
    description: str | None = None
    causes: list[str] = []
    pesticides: list[PesticideInfo] = []
    prevention: list[str] = []
    image_url: str | None = None
    refined_image_url: str | None = None
    created_at: datetime
    recapture: RecaptureGuidance | None = None


class DiagnosisSummary(BaseModel):
    id: str
    disease_name: str
    confidence: float
    crop_type: str | None = None
    image_url: str | None = None
    created_at: datetime


class PaginatedHistory(BaseModel):
    items: list[DiagnosisSummary]
    total: int
    page: int
    size: int


class DiseaseDetail(BaseModel):
    id: int
    name: str
    name_en: str | None = None
    crop_type: str
    category: str
    description: str | None = None
    causes: list[str] = []
    prevention: list[str] = []
    severity: str
    pesticides: list[PesticideInfo] = []


class HealthResponse(BaseModel):
    status: str
    model_loaded: bool
    version: str

File: frontend/src/types/diagnosis.ts
export interface PredictionItem {
  name: string;
  confidence: number;
}

export interface PesticideInfo {
  name: string;
  dilution?: string;
  usage?: string;
  active_ingredient?: string;
  safety_notes?: string;
  safety: {
    phi_days?: number;
    reentry_hours?: number;
  };
}

export type Severity = "low" | "medium" | "high";

export interface RecaptureCandidate {
  name: string;
  cue: string;
}

export interface RecaptureGuidance {
  region: string;
  instruction: string;
  margin: number;
  candidates: RecaptureCandidate[];
}

export interface DiagnoseResponse {
  id: string;
  disease_name: string;
  crop_type?: string;
  category?: "disease" | "normal";
  severity?: Severity;
  confidence: number;
  top_predictions: PredictionItem[];
  description?: string;
  causes: string[];
  pesticides: PesticideInfo[];
  prevention: string[];
  image_url?: string;
  refined_image_url?: string | null;
  created_at: string;
  recapture?: RecaptureGuidance | null;
}

export interface DiagnosisSummary {
  id: string;
  disease_name: string;
  confidence: number;
  crop_type?: string;
  image_url?: string;
  created_at: string;
}

export interface PaginatedHistory {
  items: DiagnosisSummary[];
  total: number;
  page: number;
  size: number;
}
